### PR TITLE
 [IMP] web: py_js: stop throwing errors

### PR DESCRIPTION
**Pull request by Polymorphe57**

### CODE REVIEW

**Comment by Polymorphe57** [comment]
Not important right now but could be an idea to go in this direction

File: addons/spreadsheet/static/src/helpers/model.js [hunk 1]
```diff
@@ -140,8 +141,11 @@ function containsOdooFunction(content) {
     ) {
         return false;
     }
+    const ast = parse(content);
+    if (isError(ast)) {
+        return false;
+    }
     try {
-        const ast = parse(content);
         return iterateAstNodes(ast).some(
```
Comment (Polymorphe57) [inline addons/spreadsheet/static/src/helpers/model.js:149]: try still useful?

File: addons/spreadsheet/static/src/helpers/model.js [hunk 2]
```diff
@@ -140,17 +141,16 @@ function containsOdooFunction(content) {
...
-    try {
-        const ast = parse(content);
-        return iterateAstNodes(ast).some(
-            (ast) =>
-                ast.type === "FUNCALL" &&
-                (ast.value.toUpperCase().startsWith("ODOO.") ||
-                    ast.value.toUpperCase().startsWith("_T"))
-        );
-    } catch {
+    const ast = parse(content);
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/helpers/model.js:144]: I think there's some confusion here and an over-enthusiastic grepping :smile:  This is the spreadsheet parser (which parses spreadsheet formulas and expressions), not the py parser :wink:

Comment (Polymorphe57) [reply addons/spreadsheet/static/src/helpers/model.js:144]: my bad (but the test in community pass...)

Comment (LucasLefevre) [reply addons/spreadsheet/static/src/helpers/model.js]: > but the test in community pass...

Thanks for pointing that out :) I'll add a test (https://github.com/odoo/odoo/pull/140418)

File: addons/web/static/src/search/search_model.js [hunk 3]
```diff
@@ -1576,24 +1576,13 @@ export class SearchModel extends EventBus {
...
             }
         }
-        let context;
-        try {
-            context = makeContext(contexts);
-            return context;
-        } catch (error) {
-            throw new Error(
-                _t("Failed to evaluate the context: %(context)s.\n%(error)s", {
-                    context,
```
Comment (Polymorphe57) [inline addons/web/static/src/search/search_model.js:1592]: context undefined here...
--> no information --> better just return makeContext an accept error thrown there if any

File: addons/web/static/src/core/py_js/py_interpreter.js [hunk 4]
```diff
@@ -448,14 +481,19 @@ export function evaluate(ast, context = {}) {
                 if (ast.key == "get" && typeof left === "object") {
                     return DICT[ast.key](toPyDict(left));
                 }
-                const result = left[ast.key];
+                let result;
+                try {
+                    result = left[ast.key];
```
Comment (Polymorphe57) [inline addons/web/static/src/core/py_js/py_interpreter.js:486]: hum

File: addons/web/static/src/core/error.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    };
+}
+
+/**
+ * @template a, b
+ * @param {(x: a) => b | Error} fn
+ * @param {(a | Error)[]} array
+ * @returns {(b | Error)[]}
+ */
+function _map(fn, array) {
```
Comment (Polymorphe57) [inline addons/web/static/src/core/error.js:101]: also fMap for []

File: addons/web/static/src/core/error.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+/**
+ * related to sequenceA for traversable types
+ * @template a, b
+ * @param {(x: a) => b | Error} fn
+ * @param {Record<string, a | Error>} object
+ * @returns {Record<string, b> | Error}
+ */
+function _mapObject2(fn, object) {
```
Comment (Polymorphe57) [inline addons/web/static/src/core/error.js:164]: can be defined directly without detours

File: addons/web/static/src/core/error.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+/**
+ * related to sequenceA for traversable types
+ * @template a, b
+ * @param {(x: a) => b | Error} fn
+ * @param {(a | Error)[]} array
+ * @returns {b[] | Error}
+ */
+function _map2(fn, array) {
```
Comment (Polymorphe57) [inline addons/web/static/src/core/error.js:121]: can be defined directly without detours

File: addons/web/static/src/core/py_js/py_interpreter.js [hunk 8]
```diff
@@ -422,19 +441,33 @@ export function evaluate(ast, context = {}) {
...
-                const dict = _evaluate(ast.target);
-                const key = _evaluate(ast.key);
-                return dict[key];
+                const result = evaluateSubASTs(ast);
+                if (isError(result)) {
+                    return result;
+                }
+                const { target: dict, key } = result;
+                try {
+                    return dict[key];
```
Comment (Polymorphe57) [inline addons/web/static/src/core/py_js/py_interpreter.js:450]: just check typeof dict === "object"

File: addons/web/static/src/core/error.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
+/** @odoo-module **/
+
+const ERROR = Symbol("error");
+
+/**
+ * @typedef {Object} Err
```
Comment (Polymorphe57) [inline addons/web/static/src/core/error.js:6]: @typedef {{ [typeof ERROR]: Error }} Err

**Comment by Polymorphe57** [comment]
Closed in favor of https://github.com/odoo/odoo/pull/142486 which only brings a partial solution to our original problem but is simpler and more "idomatic".

**Pull request closed without merging**